### PR TITLE
Make `AuthorizationMoyaPlugin` thread-safe

### PR DESCRIPTION
**Pull request by nuno-vieira**
## Description of the pull request

Fixes issue: https://github.com/GetStream/stream-swift/issues/40

### CODE REVIEW

File: Sources/Core/Moya/AuthorizationMoyaPlugin.swift [hunk 1]
```diff
@@ -10,22 +10,33 @@ import Foundation
...
+
     init(_ token: Token = "") {
         self.token = token
     }
-    
+
     func prepare(_ request: URLRequest, target: TargetType) -> URLRequest {
-        let token = self.token
+        var token = ""
+        serialQueue.sync {
```
Comment (angelolloqui) [inline Sources/Core/Moya/AuthorizationMoyaPlugin.swift:22]: You could use the return of the sync method like:

```
let token = serialQueue.sync { self.token }
```

**Review by angelolloqui — APPROVED** [review]
Thanks for this! Not sure if you have testing for this but maybe adding the rest for concurrency would be advisable?

File: Sources/Core/Moya/AuthorizationMoyaPlugin.swift (hunk 1 above)
Comment (nuno-vieira) [reply Sources/Core/Moya/AuthorizationMoyaPlugin.swift]: @angelolloqui Thanks for the suggestion 👍

**Comment by nuno-vieira** [comment]
> Thanks for this! Not sure if you have testing for this but maybe adding the rest for concurrency would be advisable?

What do you mean by adding the rest for concurrency?

I've tested this manually, and it is working fine. Unfortunately, our team does not have many resources allocated to this project at the moment and currently, the tests are not running it will require a bit of time to understand how the project is set up and to fix the tests.

Since we already run the tests with the thread sanitiser and this worked, it should be fine.

**Comment by angelolloqui** [comment]
Sorry, a typo. I meant adding a “test” but I understand the situation. Looking forward for this merge

**Comment by nuno-vieira** [comment]
> Sorry, a typo. I meant adding a “test” but I understand the situation. Looking forward to this merge

No problem. 👍 I've just tested this solution with the Thread Sanitizer and it is working well, with no warnings 👍 So once it is approved we will merge it.

File: Sources/Core/Moya/AuthorizationMoyaPlugin.swift [hunk 2]
```diff
@@ -10,22 +10,30 @@ import Foundation
...
         var request = request
         request.addValue("jwt", forHTTPHeaderField: "Stream-Auth-Type")
         request.addValue(token, forHTTPHeaderField: "Authorization")
-        
+
         return request
     }
+
+    func updateToken(_ token: Token) {
+        serialQueue.async {
```
Comment (polqf) [inline Sources/Core/Moya/AuthorizationMoyaPlugin.swift:35]: shouldn't this be using `.barrier`?

**Review by polqf — APPROVED** [review]

**Review by martinmitrevski — APPROVED** [review]

**Comment by angelolloqui** [comment]
I just wanted to follow up on this and say that so far looks promising! More than 50% my audience has now the updated GetStream lib and not a single crash from it received yet, compared to the many hundreds we get in a weekly basis!

**Pull request merged**